File: DAC/CPP/Day7/accountmain.cpp
#include<iostream>
#include "account.h"
using namespace std;

int main()
{

	int num;
	cout<<"Enter the size of object array:"<<endl;
	cin>>num;
	Account *ac = new Account[num];
	
	while(true)
	{
		cout << "\n-------------------Menu------------------"<< endl;
		cout<<"1.Open Account:"<<endl;
		cout<<"2.Display Account Details:"<<endl;	
		cout<<"3.Deposite Amount:"<<endl;
		cout<<"4.Withdrawl of Amount:"<<endl;
		cout<<"5.Exit:"<<endl;
		cout<<"-----------------------------------"<<endl;

		int choice;
		cout << "Enter choice :";
		cin >> choice;

		switch(choice)
		{
			case 1:
				{
				
					for(int i=0;i<num;i++)
					{
						ac[i].Accept_Details();
					}
					break;
				}
			case 2:
				{
				
					for(int i=0;i<num;i++)
					{
						ac[i].Display_Details();
					}
					break;
				}
			case 3:
				{
					cout<<"Enter the Account no You want to deposit Amount:"<<endl;
					int accno,c;
					cin>>accno;
					double amount;
					cout<<"Enter the amount you want to deposit:"<<endl;
					cin>>amount;
					for(int i=0;i<num;i++)
					{
						if(ac[i].getAccountNo() == accno)
						{
							ac[i].Deposit(amount);
							c=0;
                        				break;
                    				}
                			}

                			if (c!=0) 
					{
                    				cout << "Account Not Found!\n";
                			}	
                			break;
				}
			case 4:
				{
					cout<<"Enter the Account no You want to Withdraw Amount:"<<endl;
                                        int accno,c;
                                        cin>>accno;
                                        double amount;
                                        cout<<"Enter the amount you want to be withdrawn:"<<endl;
                                        cin>>amount;
                                        for(int i=0;i<num;i++)
                                        {
                                                if(ac[i].getAccountNo() == accno)
                                                {
                                                        ac[i].Withdrawl(amount);
                                                        c = 0;
                                                        break;
                                                }
                                        }

                                        if (c!=0) 
					{
                                        	cout << "Account Not Found!\n";
                                        }
                                        break;
				}
			case 5:
				{
					exit(0);
				}
		}
	}
		
		delete []ac;
	
	return 0;
}



/* OUTPUT
 Enter the size of object array:
2

-------------------Menu------------------
1.Open Account:
2.Display Account Details:
3.Deposite Amount:
4.Withdrawl of Amount:
5.Exit:
-----------------------------------
Enter choice :1
Enter Account Details:
Enter Account Holder Name:
Anjum
Enter Acccount Type:
Saving
Enter Balance:
1000
Enter Mobile no:
78946123
Enter Email id:
anjum@gmail.co,
Enter PAN No:
AN46785SK
Enter Account Details:
Enter Account Holder Name:
Ankita
Enter Acccount Type:
Cuurent
Enter Balance:
2000
Enter Mobile no:
879456132
Enter Email id:
ankita@gmail.com
Enter PAN No:
AK45678SK

-------------------Menu------------------
1.Open Account:
2.Display Account Details:
3.Deposite Amount:
4.Withdrawl of Amount:
5.Exit:
-----------------------------------
Enter choice :2
Account Details are:
---------------------------
Acccount No: 100
Account Holder Name: Anjum
Account Type: Saving
Account Balance: 1000
Account Holder Mobile No: 78946123
Account Holder Email: anjum@gmail.co,
Account Holder Pan No: AN46785SK
--------------------
Account Details are:
---------------------------
Acccount No: 101
Account Holder Name: Ankita
Account Type: Cuurent
Account Balance: 2000
Account Holder Mobile No: 879456132
Account Holder Email: ankita@gmail.com
Account Holder Pan No: AK45678SK
--------------------

-------------------Menu------------------
1.Open Account:
2.Display Account Details:
3.Deposite Amount:
4.Withdrawl of Amount:
5.Exit:
-----------------------------------
Enter choice :3
Enter the Account no You want to deposit Amount:
100
Enter the amount you want to deposit:
6000
Your deposit successfully done

-------------------Menu------------------
1.Open Account:
2.Display Account Details:
3.Deposite Amount:
4.Withdrawl of Amount:
5.Exit:
-----------------------------------
Enter choice :2
Account Details are:
---------------------------
Acccount No: 100
Account Holder Name: Anjum
Account Type: Saving
Account Balance: 7000
Account Holder Mobile No: 78946123
Account Holder Email: anjum@gmail.co,
Account Holder Pan No: AN46785SK
--------------------
Account Details are:
---------------------------
Acccount No: 101
Account Holder Name: Ankita
Account Type: Cuurent
Account Balance: 2000
Account Holder Mobile No: 879456132
Account Holder Email: ankita@gmail.com
Account Holder Pan No: AK45678SK
--------------------

-------------------Menu------------------
1.Open Account:
2.Display Account Details:
3.Deposite Amount:
4.Withdrawl of Amount:
5.Exit:
-----------------------------------
Enter choice :4
Enter the Account no You want to Withdraw Amount:
101
Enter the amount you want to be withdrawn:
500
Your Withdraw amount successfully

-------------------Menu------------------
1.Open Account:
2.Display Account Details:
3.Deposite Amount:
4.Withdrawl of Amount:
5.Exit:
-----------------------------------
Enter choice :2
Account Details are:
---------------------------
Acccount No: 100
Account Holder Name: Anjum
Account Type: Saving
Account Balance: 7000
Account Holder Mobile No: 78946123
Account Holder Email: anjum@gmail.co,
Account Holder Pan No: AN46785SK
--------------------
Account Details are:
---------------------------
Acccount No: 101
Account Holder Name: Ankita
Account Type: Cuurent
Account Balance: 1500
Account Holder Mobile No: 879456132
Account Holder Email: ankita@gmail.com
Account Holder Pan No: AK45678SK
--------------------

-------------------Menu------------------
1.Open Account:
2.Display Account Details:
3.Deposite Amount:
4.Withdrawl of Amount:
5.Exit:
-----------------------------------
Enter choice :5
 */

File: DAC/JAVA/Day11/src/com/cdac/acts/Account/AccountTester/AccountMain.java
package com.cdac.acts.Account.AccountTester;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.ListIterator;
import java.util.Scanner;

import com.cdac.acts.Account.Account;

public class AccountMain {

	public static void main(String[] args) {
	
		Scanner sc = new Scanner(System.in);
		DateTimeFormatter format = DateTimeFormatter.ofPattern("dd/MM/yyyy");
		List<Account> account = new ArrayList<>();
		List<Account> unmodifiableList = null;
		
		
		
		while(true)
		{
			System.out.println("\n**********AccountMenu***********");
			System.out.println("\n*******************************");
			System.out.println("1. Add Account");
			System.out.println("2. Print All Accounts");
			System.out.println("3. Deposit Amount");
			System.out.println("4. Withdraw Amount");
			System.out.println("5. Create unmodified from existing list");
			System.out.println("6. Use Upper Bound");
			System.out.println("7. Use Lower Bound");
			System.out.println("8. Exit");
			System.out.println("\n*******************************");
			
			System.out.println("\nEnter your choice");
			int choice = sc.nextInt();
			sc.nextLine();
			switch(choice)
			{
				case 1:
					try{
						
					System.out.println("Enter Account Holder Name: ");
					String name = sc.nextLine();
					System.out.println("Enter Date of Opening Account: ");
					String dop = sc.nextLine();
					LocalDate opendate = LocalDate.parse(dop, format);
					System.out.println("Enter Minimum Balance: ");
					double balance = sc.nextDouble();
					sc.nextLine();
					
					account.add(new Account(name,opendate,balance));
					}catch(IllegalArgumentException e){
						System.out.println("Error: " +e.getMessage());
					}
					break;
				
				case 2:
					System.out.println("Account Details are: ");
					System.out.println("***************************\n");
					ListIterator<Account> itr = account.listIterator();
					while(itr.hasNext())
					{
						Account ac = itr.next();
						System.out.println(ac);
					}
					System.out.println("***************************\n");
					break;
				case 3:
					System.out.println("Enter the Account No to which amount is to be depoisted");
					long acno = sc.nextLong();
					
					System.out.println("Enter the amount to be Deposited: ");
					double amount = sc.nextDouble();
					
					boolean found = false;
					
					for(Account acc: account)
					{
						if(acc.getAcno() == acno)
						{
							acc.deposit(amount);
							System.out.println("Amount Depoisted Successfully!.......New Balance is " + acc.deposit(0));
							found = true;
							break;
						}
					}
					if(!found){
						System.out.println("Account No doesnot Exists!!please enter a valid account number");
						
					}
					
					break;
				case 4:
					System.out.println("Enter the Amount from which you want Witdraw amount");
					long acno1 = sc.nextLong();
					
					System.out.println("Enter amount to be Withdraw: ");
					double draw = sc.nextDouble();
					
					boolean found1 = false;
					
					for(Account acc: account)
					{
						if(acc.getAcno() == acno1)
						{
							acc.withdraw(draw);
							System.out.println("Amount Withdraw Successfully!............NEw Balance is: " + acc.withdraw(0));
							found1 = true;
							break;
						}
					}
					if(!found1){
						System.out.println("Account No doesnot Exists!!please enter a valid account number");
						
					}
					break;
				case 5:
					if(account.isEmpty())
					{
						System.out.println("No Account Found!!!");
					}
					else
					{
						LocalDate date = LocalDate.parse("05/04/2023", format);
						unmodifiableList  = Collections.unmodifiableList(account);
						System.out.println("Unmodified list is created!!!");
						
						// this part of code creates an error of java.lang.UnsupportedOperationException as it is unmodified list
						unmodifiableList.add(new Account("Ankita",date,2000.00));
					}
					
					break;
				case 6:
					System.out.println("Using UpperBound<? extends T>.....");
					displayUpperBound(account);
					break;
					
				case 7:
					System.out.println("Using LowerBound<? super T>.......");
					displayLowerBound(account);
					break;
				case 8:
					System.out.println("Exit Successfully");
					sc.close();
					System.exit(0);
			}
			
		}
		

	}
	
	// Upper bound: You can read from this list
	private static void displayLowerBound(List<? extends Account> list) {
		for(Account ac: list)
		{
			System.out.println(ac);
		}
		
	}

	 // Lower bound: You can write to this list
	private static void displayUpperBound(List<? super Account> list) {
		for(Object obj: list)
		{
			System.out.println(obj);
		}
		
	}

}
